### DISCUSSION TITLE
Upgrade O3DE to use Qt6

### DISCUSSION THREAD

**Discussion by Shirley-yhm** [o3de/o3de#14940] [General]
I noticed that there has been some discussion about upgrading of Qt in O3DE. https://github.com/o3de/o3de/issues/11818

My team has upgraded to Qt6.3.2 on 2205 branch and it works well. Do you have any plan to upgrade O3DE to use Qt6? If yes, we'd like to contribute it to the community.

**Comment by Jackie9527** [comment]
Qt 6 was officially released on December 8, 2020.
It is a big release compared to Qt5.
It seems that upgrading applications from Qt5 to Qt6 involves lots of work to do, but don't be afraid of that because Qt officials helps developers with providing a lot docs, some of them as follows:
[What's new in Qt 6](https://doc.qt.io/qt-6/whatsnewqt6.html)
[What's New in QMetaType + QVariant](https://www.qt.io/blog/whats-new-in-qmetatype-qvariant)
[Changes to Qt Core](https://doc-snapshots.qt.io/qt6-dev/qtcore-changes-qt6.html)
[Changes to Qt Widgets](https://doc.qt.io/qt-6/widgets-changes-qt6.html)
[Qt 5 and Qt 6 compatibility](https://doc.qt.io/qt-6/cmake-qt5-and-qt6-compatibility.html)
[Qt 5 Core Compatibility APIs](https://doc.qt.io/qt-6/qtcore5-index.html)
[Porting from Qt 5 to Qt 6 using Qt5Compat library](https://www.qt.io/blog/porting-from-qt-5-to-qt-6-using-qt5compat-library)
[Qt Extras Modules in Qt 6](https://www.qt.io/blog/qt-extras-modules-in-qt-6)
[Obsolete Members for QtWin](https://doc-snapshots.qt.io/qt5-5.15/qtwin-obsolete.html)

It indeed takes time to complete the full task, but very luck, it's not that hard due to high quality of o3de codes.
Lots of bugs and security issues are been fixed by Qt6. 
Our customers and team members are benefiting from Qt6.

**Comment by Jackie9527** [comment]
@cgalvan How about upgrading to Qt6?

**Reply by cgalvan** [reply]
It's definitely come up in the past as a nice long-term goal to move to Qt6, and to vanilla Qt, it would be a big undertaking though beyond just getting it to compile. Here are my thoughts off the top of my head:

- Update the code-base to work with Qt6 on all our supported platforms (sounds like this has already been proven can be done, at least on one platform)
- Right now we are on a forked version of Qt 5.15.1 (https://github.com/o3de/qt5/tree/5.15.1-o3de) because of custom modifications to the `qtbase` and `qtimageformats` sub-modules, which can be found here https://github.com/o3de/qtbase/tree/5.15.1-o3de and here https://github.com/o3de/qtimageformats/tree/5.15.1-o3de. We would need to go through each patch one by one and determine:
    - Is this patch already fixed upstream in Qt6? If so, great!
    - If not, is there a workaround in Qt6?
    - If no workaround, do we still need the patch?
    - If we still need the patch, can we re-apply it in Qt6, or does it need to be re-implemented?
    - A number of the fixes (e.g. the High DPI changes) were actually pulled from upstream so those we can probably ignore, but some of the others will require more investigation.
- Will need to update the 3p package scripts for Qt (https://github.com/o3de/3p-package-source/tree/main/package-system/Qt) to build on all our supported platforms:
    - windows
    - linux-x86
    - linux-aarch64
    - darwin
- Update the 3p package scripts for PySide2 (https://github.com/o3de/3p-package-source/tree/main/package-system/pyside2) to build against the new Qt6 packages on all our supported platforms.
    - For Qt6, the python bindings are actually a new package called `PySide6` so we'll need to update all scripts that use the python bindings. There are likely some API changes since this is a major version bump that would require scripts to be updated.
    - PySide is used by our automated tests so the AR run should inherently catch most of any issues found while upgrading PySide
    - PySide is also used by several python tools (DCC tools, Object Tree, etc...) which would require manual testing to verify they still work as expected
- Verify our https://github.com/o3de/o3de-extras/tree/development/Gems/AzQtComponentsForPython gem still compiles against the new Qt6 and updated PySide bindings

For a change this big I would suggest revisiting after the current release has been delivered. But some of the initial investigations could begin anytime.

**Reply by Jackie9527** [reply]
Let's split and reorder the tasks, and start the journey from windows platform. 
* Prepare Qt6 binary package on windows.
  * prepare the Qt6 package.
    * select patches applied to Qt5.
    * update build scripts, make sure they work with Qt6 codes.
    * build Qt6 binary.
  * prepare the PySide6 package.
    * update build scripts, make sure they work with Qt6 codes.
    * build PySide6 binary.
  * prepare the build scripts on a special branch, named like qt6_upgrade.
* Update the code-base to work with Qt6 on windows.
  * fix compile issues related to Qt6.
  * make sure code-base works both on Qt6 and Qt5.
  * verify ui features, make sure they can work with Qt6.
  * verify python related features, make sure they can work with Qt6.
* Select the next platform which is voted by developers.

**Comment by guillaume-haerinck** [comment]
> I noticed that there has been some discussion about upgrading of Qt in O3DE. #11818
> 
> My team has upgraded to Qt6.3.2 on 2205 branch and it works well. Do you have any plan to upgrade O3DE to use Qt6? If yes, we'd like to contribute it to the community.

Reviving an old thread but QT6.8 port is indeed in progress, with no more qt custom changes

- https://github.com/o3de/o3de/issues/19081
